pkg/jsonmessage/jsonmessage.go

```diff
@@ -174,10 +175,17 @@ func cursorDown(out io.Writer, l uint) {
 // Display prints the JSONMessage to out. If isTerminal is true, it erases
 // the entire current line when displaying the progressbar. It returns an
 // error if the [JSONMessage.Error] field is non-nil.
-func (jm *JSONMessage) Display(out io.Writer, isTerminal bool) error {
+func (jm *JSONMessage) Display(ctx context.Context, out io.Writer, isTerminal bool) error {
 	if jm.Error != nil {
 		return jm.Error
 	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	default:
+	}
```

Comment:
Why check for context cancellation here? Unless we're concerned about an overly slow reader on the other end (unlikely, but possible), there's no reason to handle the context cancellation here (in `Display`) instead of just doing it in the main loop.

This is equivalent to just adding another
```go
select {
case <-ctx.Done():
	return ctx.Err()
default:
}
```

to the main loop, like:

```go
func DisplayJSONMessagesStream(ctx context.Context, in io.Reader, out io.Writer, terminalFd uintptr, isTerminal bool, auxCallback func(JSONMessage)) error {
	var (
		dec = json.NewDecoder(in)
		ids = make(map[string]uint)
	)

	for {
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}

		var diff uint
		var jm JSONMessage
		if err := dec.Decode(&jm); err != nil {
			if err == io.EOF {
				break
			}
			return err
		}
		if jm.Aux != nil {
			if auxCallback != nil {
				auxCallback(jm)
			}
			continue
		}
		if jm.Progress != nil {
			jm.Progress.terminalFd = terminalFd
		}
		if jm.ID != "" && (jm.Progress != nil || jm.ProgressMessage != "") {
			line, ok := ids[jm.ID]
			if !ok {
				// NOTE: This approach of using len(id) to
				// figure out the number of lines of history
				// only works as long as we clear the history
				// when we output something that's not
				// accounted for in the map, such as a line
				// with no ID.
				line = uint(len(ids))
				ids[jm.ID] = line
				if isTerminal {
					fmt.Fprintf(out, "\n")
				}
			}
			diff = uint(len(ids)) - line
			if isTerminal {
				cursorUp(out, diff)
			}
		} else {
			// When outputting something that isn't progress
			// output, clear the history of previous lines. We
			// don't want progress entries from some previous
			// operation to be updated (for example, pull -a
			// with multiple tags).
			ids = make(map[string]uint)
		}
		select {
		case <-ctx.Done():
			return ctx.Err()
		default:
		}
		err := jm.Display(ctx, out, isTerminal)
		if jm.ID != "" && isTerminal {
			cursorDown(out, diff)
		}
		if err != nil {
			return err
		}
	}
	return nil
}
```
But doing it after checking the message for errors.

Reply:
I would think it's a bit cleaner to have the context check happen inside the function, also wouldn't we want to have the ability to trace with OTEL into the sub-functions as well?